File: 2015/01/main.js
import * as fs from 'fs';

/**
 * @param {string} input - Puzzle input
 * @returns {number}
 */
function main(input) {
    /** @type {number} */
    let result = 0;
    /** @type {boolean} */
    let basement = false;

    for (let i = 0; i < input.length; i++) {
        if (input.charAt(i) === '(') {
            result = result + 1
        } else if (input.charAt(i) === ')') {
            result = result - 1
        }

        if (basement === false && result === -1) {
            console.log("Entered the basement at: " + (i + 1))
            basement = true
        }
    }

    return(result)
}

fs.readFile('input.txt', (err, data) => {
    if (err) throw err;
    console.log(
        "Floor: " + main(data.toString())
    );
});

File: 2015/01/main.lua
#!/usr/bin/lua

-- getData
-- @return string
local function getData()
    local puzzleInput = io.open("./input.txt", "r")

    if not puzzleInput then
        return ""
    end

    local data = puzzleInput:read("a")

    puzzleInput:close()

    return data
end

-- main
local function main()
    local data = getData()
    local answer, basement = 0, true

    for i = 1, #data do
        local char = string.sub(data, i, i)

        if char == "(" then
            answer = answer + 1
        elseif char == ")" then
            answer = answer - 1
        end

        if answer == -1 and basement then
            print("Entered the basement at: " .. i)
            basement = false
        end
    end

    print("Final floor: " .. answer)
end

main()
